### PR TITLE
[FIX] website: fix images removal on images wall snippet

### PR DESCRIPTION
**Pull request by xO-Tx**
Steps to reproduce:

- Go to website > Add an "Images Wall" block > Add images and save.
- Then try to add more images > new images are removed immediately.

Right after adding a new image to the snippet, the whole content will
be removed (see `registry.gallery` > `mode()`) and replaced with new
images in the current layout.

This adaptation will trigger a `historyStep()` on the editor side while
the "unbreakable protection mechanism" is detecting node changes inside
unbreakable ancestor and setting `_toRollback` to a truthy value,
leading to undo the last updates...

The goal of this commit is to fix this behaviour by preventing the
unbreakable protection code to be applied on image wall layout
adaptation.

In this commit, the "Layout" options are also hidden for the "Images
wall" snippet, since it has its own options to handle the layout.

opw-3240231

### CODE REVIEW

File: addons/website/static/src/js/editor/snippets.options.js [hunk 1]
```diff
@@ -3744,6 +3744,23 @@ options.registry.GridImage = options.Class.extend({
...
+    /**
+     * TODO adapt in master: used to hide the "Layout" options on "Images Wall"
+     * (which has its own options to handle the layout).
+     *
+     * @override
+     */
+    _computeVisibility() {
+        return !this.$target[0].closest('[data-snippet="s_images_wall"]');
+    },
+});
```
Comment (qsm-odoo) [inline addons/website/static/src/js/editor/snippets.options.js:3762]: Do this in its own commit, it's not related to the main fix

File: addons/website/static/src/snippets/s_image_gallery/options.js [hunk 2]
```diff
@@ -217,6 +217,9 @@ options.registry.gallery = options.Class.extend({
      * @see this.selectClass for parameters
      */
     mode: function (previewMode, widgetValue, params) {
+        if (this.options.wysiwyg) {
+            this.options.wysiwyg.odooEditor.unbreakableStepUnactive();
+        }
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_image_gallery/options.js:222]: - Are you not supposed to re-enable it at the end of the function ?
- Maybe also disable it for the smallest code as possible (so maybe not at the start of the method but before the action that is not working?
- Add a comment to justify the need for this
- Mention in the commit message that this bug is only on Chrome :)

**Review by qsm-odoo — CHANGES_REQUESTED** [review]

File: addons/website/static/src/snippets/s_image_gallery/options.js (hunk 2 above)
Comment (xO-Tx) [reply addons/website/static/src/snippets/s_image_gallery/options.js]: > * Are you not supposed to re-enable it at the end of the function ?

By checking the code on `OdooEditor`, I saw that the code of `historyStep()` should automatically enable the unbreakable check for the next one (it sets the [_checkStepUnbreakable](https://github.com/odoo/odoo/blob/16.0/addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js#L1105) and [_toRollback](https://github.com/odoo/odoo/blob/16.0/addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js#L1083) to `false`). I also checked in the codebase, and it's only disabled before the targeted action :+1:

> * Maybe also disable it for the smallest code as possible (so maybe not at the start of the method but before the action that is not working?

Moved right before `this[widgetValue]();` which applies the mode changes :+1

> * Add a comment to justify the need for this

:+1: 

> * Mention in the commit message that this bug is only on Chrome :)

:+1: 

**Remark:** while re-testing this code, I found a strange behaviour (related to Chrome too):

With chrome cache enabled, current images on `masonry` layout are not loaded correctly and stay invisible after the update (when I disable chrome cache, layout adaptation works as intended):

https://user-images.githubusercontent.com/25032876/232032275-d0bdc805-e04b-4a92-8e4f-a416d3e2cfee.mp4

The simplest fix here was to use a "clone" of the image to force the loading... (I added this in a separate commit, since not linked to the history issue :+1:)

File: addons/website/static/src/snippets/s_image_gallery/options.js [hunk 3]
```diff
@@ -222,6 +222,13 @@ options.registry.gallery = options.Class.extend({
         this.$target
             .removeClass('o_nomode o_masonry o_grid o_slideshow')
             .addClass('o_' + widgetValue);
+        // Used to prevent the editor's "unbreakable protection mechanism" from
+        // restoring Image Wall adaptations (images removed > new images added
+        // to the container & layout updates) when adding new images to the
+        // snippet.
+        if (this.options.wysiwyg) {
+            this.options.wysiwyg.odooEditor.unbreakableStepUnactive();
+        }
```
Comment (qsm-odoo) [inline addons/website/static/src/snippets/s_image_gallery/options.js:234]: I guess we can merge this to solve the OPW asap but this seems like a band-aid on the problem, I don't understand why the unbreakable system breaks things here :grimacing: Let's merge but do you have a better explanation? If not it may be worth notifying to the editor team :grimacing:

**Review by qsm-odoo — APPROVED** [review]
@robodoo rebase-ff r+

File: addons/website/static/src/snippets/s_image_gallery/options.js (hunk 3 above)
Comment (xO-Tx) [reply addons/website/static/src/snippets/s_image_gallery/options.js:234]: @qsm-odoo Yes, as I explained on discord, I couldn't find, while debugging, any suspicious code that triggers this behaviour (the code on "OdooEditor" will simply rollback the mutations of a container considered as unbreakable by checking nodes ouid...). By doing a quick check, I can see that we already had the same [issue & fix in the past](https://github.com/odoo/odoo/pull/69156) and from the discussion on the PR, it seems that "unbreakable protection mechanism" was implemented for some specific situations, but it affects the snippet changes too... :thinking: so Yes we can get more details from the editor team about this part of the code, maybe this conflict can be prevented in a global way :+1:

Comment (qsm-odoo) [reply addons/website/static/src/snippets/s_image_gallery/options.js:234]: @dmo-odoo As explained by @xO-Tx I already r+'ed such a diff in the past without really liking it. There is either something to fix in a more global way or something that I don't understand, maybe worth having a look? :)

Summary: "We apparently need to call `unbreakableStepUnactive` in some places to make it work but we don't see why the unbreakable rollback mechanism is even triggered in those places". Notice that here, the problem was also only on Chrome (not Firefox) so this is even stranger to me.
